### DISCUSSION TITLE
Strongly-Named DLL / Signed DLL

### DISCUSSION THREAD

**Discussion by muneeebsheikh** [MarkPflug/Sylvan#114] [Q&A] [answered]
Hi devs,

I am working on a .Net Framework project which requires strongly-named assembly to be referenced.
Can you provide a DLL which has Strong Name property True?

**Comment by MarkPflug** [comment, marked as answer]
No, I won't be doing that, but you can certainly pull the source code and build a signed version if you need to.

**Reply by muneeebsheikh** [reply]
Yes I need to but pulling your library and creating my own version is not the best idea, as I discussed with my team. 
But thanks anyways.

**Comment by bcalco** [comment]
@MarkPflug If I may ask, why won't you strong-sign the assemblies?

**Reply by bcalco** [reply]
Having just tried something as simple as adding a strong-name key to all the DLLs and their related test assemblies, and adding the public key of the strong-name file to the InternalsVisibleTo attribute which for some reason does not like the public key even though I followed MS's instructions to generate it, I see that it is not an easy thing to get working.  I don't fully understand all the reasons for the error messages and am frustrated because I can't get my stuff to run with Sylvan unless it is strong-signed.  There is a lot of interplay between the internals in these libraries and the test cases that seem to complicate this seemingly simple requirement a lot.

Any thoughts to get this to work appreciated. I wish I could understand in the first place why my application, which does use strong names, requires Sylvan.Data.Cvs to be strong-name signed. I am quite sure lots of DLLs I'm using aren't strong-name signed and I don't have this issue with them. The whole area is murky and the benefits of all this are even murkier.

**Reply by MarkPflug** [reply]
> why won't you strong-sign the assemblies?

I would turn that around and ask you why you need it strong named to begin with? Are you still on .NET framework and need to load it in the GAC? For me that's the only reason I've ever needed something strong named?

**Reply by bcalco** [reply]
I am on .NET Framework and building a tool that integrates with MS Office/365. I don't think the GAC is the issue per se, and frankly I don't know why I need it strong-named. I am getting errors at runtime that indicate it can't load Sylvan.Data.Csv for this reason. It's really a huge impertinence I wish I didn't have to deal with, but the Office API requires my assemblies be strong-named, and so I guess things I used must also be strong-named. However, I don't believe this to be the case of a large number of assemblies I use, so I am not sure now why I get this error message from VS and wish shit would just work, tbh.

**Reply by MarkPflug** [reply]
Are you using any other strong-named, third-party assemblies in this project? 
I'm wondering if they are in the GAC?
How is the plugin registered with MS Office? I've never looked at that API. I'm wondering if in the registration process that MS Office might be registering them in the GAC so that it can load it by name, since it wouldn't otherwise know where to load the assembly from.

**Reply by bcalco** [reply]
What's strange is, some of it does work. It is quite difficult to explain fully the context of my problem, but essentially I am integrating a scripting engine that is able to reference and load and call into the assembly. But the minute I try to access a namespace in my own assembly that also loads the same DLL, then I get the error that it can't load Sylvan.Data.Cvs "or one of its dependents" because it's not strong-named, and kaboom. But elsewhere I am able to create an instance of CsvDataReader and CsvDataReaderOptions, no problem. The issue is, when setting the StringFactory to an the StringPool.GetString method, I need to do this in C# and not the scripting engine. So I created a helper in one of my strong-named DLLs and the minute I reference a namespace in it that refers to Sylvan, kaboom. I don't even know WTH "strong-naming" supposedly achieves (it seems so easy to dance around, except in Sylvan's case, where the voodoo that you do with testing seems to stretch the utility beyond its usefulness) but now I have this stupid devops issue getting in the way of an elegant wrapper of your library that I really want to use.

**Reply by bcalco** [reply]
@MarkPflug I am using so many assemblies that I can't believe all of them are strongly named. This problem seems only to appear when one of my strongly named assemblies tries to use anything in an assembly that isn't strongly named directly. Loading it in the runtime via the script engine is not a problem, but some of the syntax of assigning the StringFactory delegate property is a bit out of the ordinary use case for interop with the CLR and so I felt it would be easier just to create a static class in one of my core assemblies that I know is already loaded. everything is good until I try to use a namespace in that DLL that refers to Sylvan (I assume if I was directly accessing another not strongly named assembly, I'd get the same burp up). What I don't get is, why I can load it indirectly, but not directly. The whole scheme seems like more pain than gain, and like swiss cheese from a security point of view (assuming that's the motivation for it).

**Reply by bcalco** [reply]
I don't have time to figure out what it would take to strongly-name Sylvan because the whole "friend assembly" BS complicates compiling any of it at all. I made all the assemblies in the project use the same snk file, and added the public key to the InternalsVisibleTo in the AseemblyInfo.cs files of Sylvan.Data and Sylvan.Common project. Nothing compiles properly, tons of errors, warnings, whatnot. When I revert to life without pointing to the snk files, all compiles no problem. I am on the verge of having to settle for an inferior alternative, and what's kind of sad is, I can't even explain why I have this stupid requirement. As far as I am concerned, Sylvan does what I want, and all I want to do is use it.

**Comment by MarkPflug** [comment]
> Loading it in the runtime via the script engine is not a problem

Can you show me an example of what that loading code looks like? That might help me understand what's causing the issue.

**Reply by bcalco** [reply]
OK so it turns out, I may have a workaround. Thankfully, your StringFactory delegate type only requires known types for input/output, which means I can assign Func<char[],int,int,string> without any reference to the Sylvan (or Ben.StringIntern) libraries in my C# wrapper code. The language I'm integrating has a function type called "Callable" so all the indirection I need to make this work is achieved with the following C# code:

```csharp
    public static Func<char [ ], int, int, string> GetStringFactoryFromCallable ( Callable aCallable )
    {
      return ( char [ ] bytes, int offset, int length ) =>
      {
        return ( string ) aCallable.Call ( bytes, offset, length );
      };
    }
```

I haven't tested it yet, but it seems I can simply wrap calling StringPool.GetString() and InternPool.Intern(char[]) in the scripting language via direct .NET interop. 

I have no idea how all this strong-name signing business is supposed to make anything more secure, especially now that I know how to get around it. I remember now I had an issue with another library, whose author decided to support strong naming. The situation appears to be that as long as my strongly-named library doesn't directly load one that isn't strongly named, I can use this kind of indirect loading of "unclean" assemblies. As soon as one of the types in them needs to be surfaced in my code, it will go kaboom and complain about the lack of strong-name signing. In the case of that library there was no avoiding putting a library-defined type in the delegate signature, but in his case it was easy to add strong-name signing and he didn't mind doing it.

**Reply by MarkPflug** [reply]
So, it sounds like you have a working solution?

I don't think strong naming was actually meant to be a security feature. It was more about ensuring that a library can be given a unique name, without there having to be a global registry.  We have such a global registry now though, as nuget.org provides a registry for unique library names. Nuget prevents someone else from uploading a package named "Csv", because someone already created such a package. So, in the scope of nuget names, that unambiguously names the "Csv" library by Steve Hansen. If you had two dlls named csv.dll, but they were written by different people, and they were both v1.0.0, how could you differentiate them? That's the problem that strong naming was meant to solve. At least that's my understanding. If each library had a strong name, you could use that to differentiate.

**Reply by bcalco** [reply]
That makes sense - in any case, yeah, I'm able to produce a delegate (albeit kind of indirectly) from the language that StringFactory should be able to get on with, without triggering the wrath of the strong-name signing gods.  If I run into any problems that look like they are Sylan related, I'll let you know.

**Reply by bcalco** [reply]
It seems I spoke too fast - I am having problems setting the result of that call, which is a Func<char[],int,int,string>, to the value of CsvDataReaderOptions.StringFactory ... but it appears to be a language issue. From your POV, that should be compatible with the StringFactory delegate type, right?

**Reply by MarkPflug** [reply]
You might try this...
If all you are using from Sylvan.Common is the string pool type, just copy that code into your project and reference it from there. The beauty of open source. It really isn't that much code, and you shouldn't ever have to mess with it.

**Reply by MarkPflug** [reply]
This is the file:
https://github.com/MarkPflug/Sylvan/blob/main/source/Sylvan.Common/StringPool.cs
Not even 200 lines of code, entirely self-contained, you can just drop that file into your project with no changes.

**Reply by bcalco** [reply]
I may have to go that route while the script language issue(s) get worked out for this. Ben.StringIntern is strong-name signed so I can also use that more directly. Looks like I just need to create some C# wrapper classes with simpler interfaces around the assignment of the StringFactory property.

**Reply by bcalco** [reply]
@MarkPflug  Just to give you a quick update - I was able to resolve my actual issue at the script language level, dealing with the assignment of delegate types. However, the strong-name signing problem I ran into by attempted to do that assignment in C# appears to be solvable by loading unsigned assemblies *indirectly* through the scripting engine, and avoiding direct use by my strong-name signed assemblies. I don't know why that works, but it does. That is, I was able to write a C# DLL, not strong-name signed, that my strong-name signed project *referred* to but did not actually use (the purpose being to make sure it is copied to the target directory). Loading that DLL from the script side allows me to refer to types in Sylvan (which it uses directly) no problem. But it turned out being able to assign lambdas directly to the delegate property in my script language meant I didn't have to set the stringfactory property in a C# wrapper. But when / if I ever need to, now I know how to avoid the problem I ran into here, and I even have a placeholder for it.
